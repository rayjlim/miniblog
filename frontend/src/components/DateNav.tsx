import useDateNav from '../hooks/useDateNav';

const DateNav = ({ updateDate, date }: { updateDate: (date: string) => void, date: string }) => {
  const { inputDate, handleButtonDirection, updateDateInput } = useDateNav(updateDate, date);

  return (
    <div className="grid-3mw container">
      <button
        onClick={e => handleButtonDirection(e)}
        className="btn btn-info btn-lrg"
        value="-1"
        id="prevBtn"
        type="button"
        title="alt + comma"
      >
        <i className="fa fa-chevron-left" />
        Prev
      </button>
      <div>
        <input
          type="text"
          className="form-control"
          id="formDpInput"
          onChange={e => updateDateInput(e)}
          ref={inputDate}
        />
      </div>

      <button
        onClick={e => handleButtonDirection(e)}
        className="btn btn-success btn-lrg"
        value="1"
        id="nextBtn"
        type="button"
        title="alt + period"
      >
        Next
        <i className="fa fa-chevron-right" />
      </button>
      <button
        onClick={e => handleButtonDirection(e)}
        className="btn btn-warning btn-lrg"
        value="today"
        type="button"
      >
        Today
      </button>
    </div>);
};

export default DateNav;
