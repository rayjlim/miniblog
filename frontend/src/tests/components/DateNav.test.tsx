import { render, screen, fireEvent } from "@testing-library/react";
import renderer from 'react-test-renderer';

import { vi } from 'vitest';
import jest from 'jest-mock';
import '@testing-library/jest-dom';

import DateNav from "../../components/DateNav";

const fetchMock = vi.fn(() => ({
  json: vi.fn(() => Promise.resolve({ uploadDirs: ['dir1'], dirContent: ['media1'] })),
  ok: true,
}));

vi.stubGlobal('fetch', fetchMock);
const RealDate = Date.now
let spy: any = null;

console.log = () => {};

describe("DateNav component", () => {
  beforeAll(() => {
    console.log('Normal:   ', new Date().getTime())

    const mockDate = new Date(Date.parse('2019-04-22'));
    spy = jest
      .spyOn(global, 'Date')
      .mockImplementation(() => mockDate);
    global.Date.now = jest.fn(() => new Date('2019-04-22T00:00:00Z').getTime())
    global.Date.UTC = jest.fn(() => new Date('2019-04-22T00:00:00Z').getTime())
    console.log('Mocked:   ', new Date().getTime())
  })

  afterAll(() => {
    spy?.mockRestore()

    console.log('Restored: ', new Date().getTime())
    global.Date.now = RealDate;
  })

  afterEach(() => {
    vi.restoreAllMocks();
  })

  it("should render DateNav component correctly", () => {
    const mockCallBack = jest.fn();
    render(<DateNav updateDate={mockCallBack} date={'2023-11-12'} />);
    const btn = screen.getByText(/Prev/);
    expect(btn).not.toBeNull();
    expect(btn).toBeInTheDocument();
    const btn2 = screen.getByText(/Next/);
    expect(btn2).toBeInTheDocument();
    const btn3 = screen.getByText(/Today/);
    expect(btn3).toBeInTheDocument();
  });

  it("should render DateNav component correctly", () => {
    const mockCallBack = jest.fn();
    const tree = renderer
    .create(<DateNav updateDate={mockCallBack} date={'2023-11-12'} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
  });

  it("should go to Today on btn click", () => {
    const mockCallBack = jest.fn();
    render(<DateNav updateDate={mockCallBack} date={'2023-11-08'} />);

    const input = screen.getByLabelText('date-input');

    expect(input).toBeInTheDocument();
    fireEvent.change(input, {target: {value: '2022-01-01'}});

    const btn = screen.getByText(/Today/);
    expect(btn).toBeInTheDocument();
    fireEvent.click(btn);

    expect(mockCallBack).toHaveBeenCalledTimes(2);
    expect(mockCallBack.mock.calls[0][0]).toBe('2022-01-01');
    expect(mockCallBack.mock.calls[1][0]).toBe('2019-04-22');
  });

  it("should go to Prev on btn click", () => {
    const mockCallBack = jest.fn();
    render(<DateNav updateDate={mockCallBack} date={'2023-11-10'} />);

    const btn = screen.getByText(/Prev/);
    expect(btn).toBeInTheDocument();
    fireEvent.click(btn);

    expect(mockCallBack).toHaveBeenCalledTimes(1);
    expect(mockCallBack.mock.calls[0][0]).toBe('2023-11-09');
  });

  it("should go to Next on btn click", () => {
    const mockCallBack = jest.fn();
    render(<DateNav updateDate={mockCallBack} date={'2023-11-10'} />);

    const btn = screen.getByText(/Next/);
    expect(btn).toBeInTheDocument();
    fireEvent.click(btn);

    expect(mockCallBack).toHaveBeenCalledTimes(1);
    expect(mockCallBack.mock.calls[0][0]).toBe('2023-11-11');
  });
  it("should go to new date on input select", () => {
    const mockCallBack = jest.fn();
    render(<DateNav updateDate={mockCallBack} date={'2023-11-10'} />);

    const input = screen.getByLabelText('date-input');

    expect(input).toBeInTheDocument();
    fireEvent.change(input, {target: {value: '2022-01-01'}});

    expect(mockCallBack).toHaveBeenCalledTimes(1);
    expect(mockCallBack.mock.calls[0][0]).toBe('2022-01-01');
  });
});
