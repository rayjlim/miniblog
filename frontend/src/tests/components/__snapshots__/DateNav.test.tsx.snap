// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateNav component > should render DateNav component correctly 1`] = `
[
  <div
    className="grid-3mw container col-3"
  >
    <button
      className="btn btn-info btn-lrg success"
      id="prevBtn"
      onClick={[Function]}
      title="alt + comma"
      type="button"
      value="-1"
    >
      <i
        className="fa fa-chevron-left"
      />
      Prev
    </button>
    <div>
      <input
        aria-label="date-input"
        className="form-control"
        onChange={[Function]}
        type="date"
      />
    </div>
    <button
      className="btn btn-success btn-lrg attention"
      id="nextBtn"
      onClick={[Function]}
      title="alt + period"
      type="button"
      value="1"
    >
      Next
      <i
        className="fa fa-chevron-right"
      />
    </button>
  </div>,
  <button
    className="btn btn-warning btn-lrg"
    onClick={[Function]}
    type="button"
    value="today"
  >
    Today
  </button>,
]
`;
